File: 백준/Silver/1012. 유기농 배추/유기농 배추.js
const fs = require('fs');
const input = fs.readFileSync(0, 'utf-8').toString().trim().split('\n').map(x => x.split(' ').map(Number));
const [T] = input[0];
let index = 1;
let answer = '';

while (index < input.length) {
    const [M, N, K] = input[index++];
    const cabbage_positions = input.slice(index, index + K);
    const field = Array.from(new Array(N), () => new Array(M).fill(0));
    let count = 0;
    
    for (let [m, n] of cabbage_positions) {
        field[n][m] = 1;
    }
    
    for (let n = 0; n < N; n++) {
        for (let m = 0; m < M; m++) {
            if (field[n][m] === 1) {
                countCabbages(n, m);
                count++;
            }
        }
    }
    
    answer += `${count}\n`;
    index = index + K;
    
    function countCabbages(n, m) {
        field[n][m] = 0;
    
        const dn = [-1, 0, 1, 0];
        const dm = [0, 1, 0, -1];
    
        for (let i = 0; i < 4; i++) {
            const nn = n + dn[i];
            const nm = m + dm[i];
        
            if (nn < 0 || nn >= N) continue;
            if (nm < 0 || nm >= M) continue;
            if (field[nn][nm] === 0) continue;
            
            countCabbages(nn, nm);
        }
    }
}

console.log(answer);

File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
T = int(input())

for _ in range(T):
    M, N, K = map(int, input().split())
    F = [[0 for m in range(M)] for n in range(N)]
    visited = [[False for _ in range(M)] for _ in range(N)]
    di = [0, 1, 0, -1]
    dj = [1, 0, -1, 0]
    count = 0
    
    for k in range(K):
        x, y = map(int, input().split())
        F[y][x] = 1
        
    for i in range(N):
        for j in range(M):
            if F[i][j] == 0 or visited[i][j]:
                continue
                
            count += 1
            stack = [[i, j]]
            
            while len(stack) > 0:
                ci, cj = stack.pop()
                visited[ci][cj] = True
                
                for k in range(4):
                    ni = ci + di[k]
                    nj = cj + dj[k]

                    if (ni >= 0 and nj >= 0 and ni < N and nj < M
                        and not visited[ni][nj] and F[ni][nj] == 1):
                        stack.append([ni, nj])
                        
    print(count)